Add tutorial for executor design (#4361)
PointCloudLibrary/pcl: doc/advanced/content/executor_design.rst

.. _executor_design:

Executor Design
------------------------------

In this document we discuss about executors which are part of C++ standard committee proposals

> and their relevance, design as well as implementation in PCL.

Reviewer: weird placement of the word `and`

Author: Reworded, PTAL

An agent is bound to an execution context, and hence to one or more of the resources that the context
represents.
Typical examples of an execution context are a thread pool or a runtime environment provided by a GPU.

**Execution Agent**

It is a unit of execution of a specific execution context that is mapped to a single invocation
of a callable function on an execution resource. An execution agent can too have different semantics which
are derived from the execution context.
Typical examples of an execution agent are a CPU thread or GPU execution unit.

**Executor**

An executor is an object associated with a specific execution context. It provides one or more execution
functions for creating execution agents from a callable function object. The execution agents created are
bound to the executor’s context, and hence to one or more of the resources that context represents.

Why does PCL need executors?
=================================

PCL has a a diverse collections of modules with various algorithms. Many of these implementations of algorithms
have a diverse set of algorithms targeting various facilities such as SIMD, OpenMP, GPU (CUDA) etc.

> Since each facility has a different set of API's and interface, so they have separate implementations,

Reviewer: APIs.

Also your sentence reads as
> has a different set of Application Programming Interfaces and interface

you might want to reword that.

> some of them are in the form of separate classes:.

Reviewer: Stray colon .

Some of the problems with the current scenario are:

1. **Divergent Implementation**

> The distinct implementations of algorithms leads to disparity in the codebase over time. Since the more popular
> implementation get bug fixes, new features and undergoes refactoring while the other implementations remain
> untouched.

Reviewer: ```suggestion
The distinct implementations of algorithms leads to disparity in the codebase over time. The more popular
implementation gets bug fixes, new features and undergoes refactoring, while the other implementations remain
untouched.
```

Example: The parallelized version of an algorithm might be more popular so it will get bug fixes overtime while
those bugs continue to persist in the serial implementations of that algorithm.

2. **Non Uniform API**
The API for one of the implementations for a specific facility might undergo some changes to accommodate
the some of the new features utilizing some of the tools provide by the facilities optimizations or

> to adapt to the interface provided by the facilities.

Reviewer: please revise this. The middle of the sentence is confusing. 
> tools provides by the facilities optimization

:/

Author: Reworded, PTAL

Example: Parallel implementations expose API's to allow configuring the degree of parallelism which is completely
absent from sequential implementations.

3. **Inextensible Design**
The current design doesn't support using new facilities like thread pools, multi-gpu support or nesting
facilities. To add support for these facilities completely new implementations will have to be written
for every algorithm.
Example: It isn't possible to run vectorized code (SIMD) while running a parallel implementation which
uses OpenMP.

4. **Code Duplication**
Even if different facilities might require slightly different implementations, a lot of the code can
be shared. Having different implementations just leads to a majority of the code base being duplicated

> and only some of the code gets modified inorder to adapt to the interface provided by the facility.

Reviewer: in order

> Example: Most of OpenMP code is quite similar to the sequential implementation with only some additions.

Reviewer: Most OpenMP code

So having separate classed for OpenMP implementations is quite redundant.

5. **Maintenance Overhead**
Maintaining several implementations of the same algorithm is a labour and time intensive task.
Porting changes from one implementation to another and propagating bug fixes to all the implementations
is time consuming. Lack of time to propagate the changes will lead to divergent implementations as mentioned
above.

Design Considerations
=================================

The executor design proposal for C++, aims to build a generalized and extensible framework surrounding

> executors. This is necessary since it needs to support a wide rage of uses cases so as to attempt to cater

Reviewer: range (I hope)

everyone's needs, in order for it to be accepted into the standard library.

In PCL, the need for executors is limited to certain features and there is no need for the entire feature set
as mentioned in the proposal. Concepts such asynchronous operations and task based parallelism are not present
or needed in PCL at the moment so creating a design which incorporates all those features would be unnecessary.
The main use cases in PCL are:

1. Provide a uniform API for executing existing algorithms on different facilities giving users the freedom
   to switch between facilities with ease.

2. Reducing code duplication and trying to avoid completely different implementations of the same algorithm

> 3. Provide a simple and easy to use mechanism to customize the execution context which users can also access

Reviewer: needs commas

> 4. Expose some of the underlying features offered by the the various facilities in a standardized manner

Reviewer: duplicate the

> 5. Provide a mechanism to automatically choose the best facility to run an algorithm in case the user does
>    not explicitly specify which facility to use

Reviewer: ```suggestion
5. Provide a default, implicit mechanism to automatically choose the best facility to run an algorithm.
```

6. Be extensible enough to allows users to specify their own executors or customize the ones provided by PCL

> 7. Ensure as little overhead as possible, if possible make everything compile time. This is also ensure all

Reviewer: new line missing
```suggestion
7. Minimize runtime overhead by exploiting compile time resolution whenever appropriate. This is also ensures 
```

>    errors are caught at compile time, providing guarantee at compile time that an executor would work.

Reviewer: too much repetition of the expression compile time

Author: Reworded

8. Last but not the least be forward compatible with the upcoming executor design so that PCL is compatible with

>    them when they become a part of the standard specification.

Reviewer: compatible with **it** when **it** becomes

Accepted Design
=================================

> The currently implemented design draws heavy influence from some of the current implementations which are being

Reviewer: The current

Too much repetition of the word stem `current`

developed in light of the proposal which are:

* `executors-impl <https://github.com/executors/executors-impl>`_

* `cudex <https://github.com/jaredhoberock/cudex>`_

The two main elements of this design are namely executors themselves and  executor properties.

**Executor Design**

> So as per the proposal the technical definition of an executor is:

Reviewer: redudant so

An Executor should be a CopyConstructible and EqualityComparable type that provides a function named

> execute that eagerly submits work on a single execution agent created for it by the executor.

Reviewer: highlight important terms

There are two available execution functions in any executor:

1. `execute`
It takes a nullary callable (a callable which takes no arguments and returns void) and executes
the callable on a single execution agent exactly once.

2. `bulk_execute`
It takes a callable (which returns void but takes an index parameter as argument) and a shape which
corresponds to the number of invocations of the callable. This function generates execution agents
equal to the number of invocations in bulk and then each execution agent invokes the callable once.
The index of the execution agent is passed as argument to the callable, so that the callable
knows the invocation index.

The difference between simply calling execute repeatedly and bulk_execute is that bulk_execute
leverages the facilities API to generate execution agents in bulk which is more efficient then creating

> them one by one.

Reviewer: Needs Kunal's cuda example

Author: Added. PTAL.

How these executors call the callable internally is dependent on the implementation of each executor
and some aspects of the execution can be customized through properties.

The index passed in bulk execute can be used to internally partition certain parts of the code to only run

> on on specific indexes.

Reviewer: double on

Example: It can be used to split the iteration of a loop between the execution agents.

**Shape and Index**

The shape and index will vary depending on the facilities, so a mechanism has been provided to customize
their types. By default in they are `std::size_t`.

The shape or index can be specified by a type or an alias for a type inside the executor with the names

> `shape_type` and `executor_type`. There also exists type traits namely `executor_shape` and `executor_index`

Reviewer: shape_type and index_type?

> to access the type the executors shape or index.

Reviewer: ```suggestion
to access the type of the executor's shape or index.
```

**Executor Properties**
Executor properties are objects associated with an executor. They are used to customize various aspects
of the executor related to execution and are also used to provide guarantees.

> The properties which are implemented in PCL currently are:

Reviewer: which are currently implemented

* Blocking

This specifies whether or not execution inside an execution function should wait/block till all
execution agents are done executing. There are 3 mutually exclusive blocking properties

> `blocking.always`, `blocking.never` and `blocking.possibly` their role can determined by their names

Reviewer: lack of full stop after blocking types are enumerated.

itself. The default is `blocking.always`.

* Allocators

> It specified the allocator,to associate with an executor. A user may use this

Reviewer: specifies?

> property to suggest the use the specified preferred allocator when allocating storage necessary

Reviewer: to much use :drum: of the word stem `use` in this sentence

for execution. The default is the specialization `allocator_t<void>` which indicates to
use the default allocator available in the system.

As of now only these 2 properties are supported in PCL but even they are not fully supported by
the provided executors.

> It is compulsory for a property to define a default property, which indicates the the property

Reviewer: duplicate the

value even if an executor doesn't explicitly support that property.

**Property Customization Mechanism**

Properties of an executor are specified using the template parameters of an executor class template.
A user may introduce a new property to an executor by defining a property type and
specializing either the `require` or `prefer` and `query` member functions inside the executor.

> The properties of an executor can strongly or weakly associate properties which are supported by an

Reviewer 2: Reword this?

Author: PTAL

* Tag Dispatching

This was was one of the initial design considerations. It seemed like an attractive choice due to simplicity,
could choose the allowed overload resolution to choose the best option at compile time by allowing
tags to be inherited. These tags would serve as placeholders till executors were standardized.
They lacked all the features of executors and offered no customization. Once executors were even standardized
large parts of the codebase would need to be refactored to support executors.

* Base Executor

The idea was to have single base executor from which all executors would derive from. This would allow
all the common cod to be shared among all the executors. The base executors has a CRTP based design which
is used to access the properties of the derived executor. It also allowed simplification in many areas
such as being able to reference any derived executor. Basic properties of all executors like
copy constructors, overloaded equality operators.
The CRTP mechanism had some restrictions in the sense that there was still a need for templates and
their were concerns regarding explicitly passing all the properties to the base class using CRTP as

> it was felt to be unnecessary. Having inheritance also introduces runtime polymorphism which was against

Reviewer 2: Expand on why CRTP needed runtime polymorphism since CRTP is known to be the compile-time alternative of virtual

Author: Added some info. PTAL.

one of the design consideration i.e. have everything compile time and avoid any overhead.

> The base executor besides a few common functionality didn't do much and was discarded.

Reviewer 2: The base executor didn't add a lot besides allowing code sharing of a few common functionality and thus it was discarded

* Overload Mechanism

> In this design, the supported executor properties specified as templated parameters were inherited from.

Reviewer 2: Reword

Author: PTAL

> With this design executors would not need to provide the `require` or `prefer` and `query` member functions

Reviewer 2: With this design, executors

comma added

> inside the executor for each property it wants to allows association with and support customizing.

Reviewer 2: allow

no plural

This design was not fully explore, so the potential drawbacks are not know completely. A snippet
is available `here <https://godbolt.org/z/zhKM6e>`_. It was decided to not go with this design forward
due to the potential complications, minimal support of properties for executors in PCL and deviation
from the property customization mechanism design implemented in other implementations of the executor design
proposal mentioned above in the document. Once more properties to be used w.r.t PCL are made available and
use cases for these properties comes up. This design can be considered and looked at.

> and

Reviewer 2: Randomly added `and`

* Unified Shape with Compile Time Support
Since each executor could have a different shape type, a unified shape type would provide a more uniform
API. The compile time support for shapes would also bring performance benefits. The compile time aspect
was ditched since in most scenarios the value of the shape was determined at runtime and rarely at compile

> time. The unified shape was also deemed unnecessary since for most executors `std::size_t` or and `int` was

Reviewer 2: or and?

> sufficient and only a custom shape was needed for GPU (CUDA) code for which tje shape could explicitly

Reviewer 2: a custom shape was needed only for GPU

for which the shape could be explicitly

specified.

Conclusion
=================================

> Executors in PCL a very new concept and will mature as more functions are supported and is more widely used.

Reviewer 2: Reword, break into 2 sentences if needed.

Suggestion: will mature as they are used more widely across PCL

Author: I feel a single statement will be a better fit. Slightly changed the wording as suggested.

> The design will also go though more changes and iterations as more feedback from the community is received

Reviewer 2: feedback. No `more`

.. _executor_design:

Executor Design
------------------------------

> In this document we discuss about executors (which are part of C++ standard committee proposals)

Reviewer 2: ```suggestion
In this document we discuss about executors (which are part of C++ standard committee proposals),
```

everyone's needs, in order for it to be accepted into the standard library.

In PCL, the need for executors is limited to certain features and there is no need for the entire feature set
as mentioned in the proposal. Concepts such asynchronous operations and task based parallelism are not present
or needed in PCL at the moment so creating a design which incorporates all those features would be unnecessary.
The main use cases in PCL are:

1. Provide a uniform API for executing existing algorithms on different facilities giving users the freedom
   to switch between facilities with ease.

2. Reducing code duplication by trying to avoid completely different implementations of the same algorithm

3. Provide a simple and easy to use mechanism to customize the execution context, which users can also access.

4. Expose some of the underlying features offered by the various facilities in a standardized manner.

5. Provide a default, implicit mechanism to automatically choose the best facility to run an algorithm.

6. Be extensible enough to allows users to specify their own executors or customize the ones provided by PCL

> 7. Minimize runtime overhead by exploiting compile time resolution whenever appropriate. This is also ensure all

Reviewer: this also ensures errors...

(all errors is an overstatement)

   errors are caught at compile time, providing guarantee an executor would work if the code compiles successfully.

8. Last but not the least be forward compatible with the upcoming executor design so that PCL is compatible with
   it when it becomes a part of the standard specification.

Accepted Design
=================================

> The implemented design in PCL draws heavy influence from some of the implementations which are being
> developed in light of the proposal which are:

Reviewer 2: ```suggestion
The implemented design in PCL draws heavy influence from the following in-development implementations:
```
Just a suggestion

**Executor Design**

As per the proposal the technical definition of an executor is:
An executor should be a `CopyConstructible` and `EqualityComparable` type that provides a function named
`execute` that eagerly submits work on a single execution agent created for it by the executor.

There are two available execution functions in any executor:

1. `execute`
It takes a nullary callable (a callable which takes no arguments and returns void) and executes
the callable on a single execution agent exactly once.

2. `bulk_execute`
It takes a callable (which returns void but takes an index parameter as argument) and a shape which
corresponds to the number of invocations of the callable. This function generates execution agents
equal to the number of invocations in bulk and then each execution agent invokes the callable once.
The index of the execution agent is passed as argument to the callable, so that the callable
knows the invocation index.

The difference between simply calling execute repeatedly and bulk_execute is that bulk_execute
leverages the API of the facility to generate execution agents in bulk which is more efficient then creating

> them one by one.

Reviewer: revise for commas

      template <typename F>
      void execute(F&& f) const {
        f();
      }

      // bulk_execute invokes the callable n times and passes the index of execution agent
      // In case of inline_executor each execution agent is mapped to the same thread is invoked sequentially
      template <typename F, typename... Args>
      void bulk_execute(F&& f, const shape_type& n) const {
        for (shape_type index = 0; index <  n; ++index)
          f(index);
      }

      // The query customization point being specialized to show that the executor supports
      // the blocking.always property
      static constexpr auto query(const blocking_t&) noexcept {
        return blocking_t::always;
      }

      // The require customization point being specialized to allow strongly associating the executor with
      // the blocking.always property
      inline_executor<blocking_t::always_t, ProtoAllocator> require(const blocking_t::always_t&) const {
        return {};
      }
    };

Best Fit Executor
=================================

In most scenarios users of PCL do not care much about performance and simply want their code to run

> fast without any additional steps performed from their side. For this reason their was a need for

Reviewer: ... there was a need...

> mechanism called best fit, which automatically selects best possible executor and its customizes its

Reviewer: a mechanism...

duplicate its

property in order to give the user good performance out of the box based on their system configurations
and the function they are calling.

> The mechanism works be choosing an executor based on the availability of an executor (depends on

Reviewer: by choosing

hardware/software of a system) and the priority of executors specified in function by PCL

> maintainers & contributors which will give good performance. The executor properties can also be

Reviewer: the work executor appears three times in this sentence.

> customized on for a better fit for certain scenarios. Besides this runtime checks are also be specified

Reviewer: Executor properties can also be customized to better fit certain scenarios.

and on the basis of these runtime checks the executors are further filtered to selected most appropriate

> one. The two supported mechanisms in PCL currently are `enable_exec_with_priority` and

Reviewer: Revise this sentence for commas, grammar and to reduce the repetition of the expression "runtime checks".

> `enable_exec_on_desc_priority`. You can read more about them in the code API.

Reviewer: Currently there are...

Alternative Designs Considered
=================================

There has been a lot of deliberation and discussion regarding all the design aspects of executors in PCL.
The design has gone through multiple iterations before a consensus was reached. Some of the major alternate
design proposals that were rejected were:

* Tag Dispatching

This was was one of the initial design considerations. It seemed like an attractive choice due to simplicity,

> could choose the allowed overload resolution to choose the best option at compile time by allowing

Reviewer: due to its simplicity, allowing overload resolution to choose the best options at compile time.

tags to be inherited. These tags would serve as placeholders till executors were standardized.
They lacked all the features of executors and offered no customization. Once executors were even standardized
large parts of the codebase would need to be refactored to support executors.

* Base Executor

> The idea was to have single base executor from which all executors would derive from. This would allow

Reviewer: ... to have a single ....

> all the common cod to be shared among all the executors. The base executors has a CRTP based design which

Reviewer: code

The base executor ...

> is used to access the properties of the derived executor. It also allowed simplification in many areas

Reviewer: In the previous sentence you used the present tense to refer to the base executor. Now you refer to it in the past. Be uniform.

Execution created through different facilities has different synchronization properties.
For example, an OpenMP parallel for loop is synchronous because the spawning thread blocks until the loop
is complete due to an implicit barrier at the end of the parallel region by default. In contrast, the execution
of GPU kernels is typically asynchronous; kernel launches return immediately, and the launching thread
continues its execution without waiting for the kernel's execution to complete. Work submitted to a thread

> pool may or may not block the submitting thread. The correct code must account for these synchronization

Reviewer: questionable change

Reviewer 2: `A` correct code?

Reviewer: I would stick to the original in this case. Simply

Correct code must account...

**Execution Agent**

It is a unit of execution of a specific execution context that is mapped to a single invocation
of a callable function on an execution resource. An execution agent can have different semantics that

> is derived from the execution context.

Reviewer: this should be plural. the semantics are derived (plural) not the execution agent (singular)

functions for creating execution agents from a callable function object. The execution agents created are
bound to the executor's context, and hence to one or more of the resources that context represents.

Why does PCL need executors?
=================================

PCL has a diverse collection of modules with various algorithms. Many of these implementations target
a diverse set of facilities such as SIMD, OpenMP, GPU, etc. Since each facility has a unique set
of interfaces that are often coupled with low-level implementation details, so some of them are required
to have separate implementations i.e., separate classes.
The current implementation suffers from a few drawbacks, such as:

1. **Divergent Implementation**
The distinct implementations of algorithms lead to disparity in the codebase over time. The more popular
implementation gets bug fixes, new features and undergoes refactoring while the other implementations remain
untouched.
Example: The parallelized version of an algorithm might be more popular so it will get bug fixes overtime while
those bugs continue to persist in the sequential implementation of that algorithm.

> 2. ** Non-Uniform API**

Reviewer: extra leading space

=================================

In most scenarios users of PCL do not care much about performance and simply want their code to run
fast without any additional steps performed from their side. For this reason, there was a need for a
a mechanism called the best fit, which automatically selects the best possible executor and customizes its
property in order to give the user good performance out of the box based on their system configurations
and the function they are calling.

The mechanism works by choosing an executor based on its availability (depends on
hardware/software of a system) and its priority specified in function by PCL
maintainers & contributors, which will give a good performance. Executor properties can also be customized
to better fit certain scenarios. Besides these runtime checks are also specified, and on the basis
of these runtime checks, the executors are further filtered to select the most appropriate

> one. The two supported mechanisms in PCL currently there are `enable_exec_with_priority` and

Reviewer: This last sentence still reads weird. Currently there are two supported mechanisms...

Author: Reworded

`enable_exec_on_desc_priority`. You can read more about them in the code API.

Alternative Designs Considered
=================================

There has been a lot of deliberation and discussion regarding all the design aspects of executors in PCL.

> The design had gone through multiple iterations before a consensus was reached. Some of the major alternate

Reviewer: The design went...

design proposals that were rejected were:

* Tag Dispatching

This was one of the initial design considerations. It seemed like an attractive choice due to its simplicity,
allowing overload resolution to choose the best options at compile time. This was achieved by allowing
tags to be inherited. These tags would serve as placeholders till executors were standardized.
They lacked all the features of executors and offered no customization. Once executors were even standardized
large parts of the codebase would need to be refactored to support executors.

* Base Executor

The idea was to have a single base executor from which all executors would derive from. This would have allowed
all the common code to be shared among all the executors. The base executor had a CRTP based design that

> was used to access the properties of the derived executor. It also allowed simplification in many areas, such as being able to reference any derived executor. Basic properties of all executors like

Reviewer: simplification**s**

there were concerns regarding explicitly passing all the properties to the base class using CRTP as
it was felt to be unnecessary. Having inheritance also introduces runtime polymorphism as the derived executor
would override some of the base executors' method, and the call to the overridden function is resolved at run time.
This is opposed to one of the design consideration i.e., have everything compile-time and avoid any overhead.
The base executor didn't add a lot besides allowing code sharing of a few common functionalities and thus
it was discarded.

* Property Inheritance

In this design, the executor inherits from the properties it supports (which are defined as template parameters
for the executor class). With this design, executors would not need to provide the `require` or `prefer`
and `query` member functions inside the executor for each property it wants to allow association with and
support customizing. This design was not fully explored, so the potential drawbacks are not known completely.
A snippet is available `here <https://godbolt.org/z/zhKM6e>`_. It was decided not to go forward with this design

> due to the potential complications, minimal support of properties for executors in PCL, and deviation

Reviewer: potential complications doesn't really add any info to this sentence

Author: Removed and reworded

from the property customization mechanism design implemented in other implementations of the executor design
proposal mentioned above in the document. Once more properties to be used, w.r.t PCL is made available, and

> use cases for these properties come up. This design can be considered and looked at a future stage.

Reviewer: Sentence still reads very weird.

Author: Reworded

The idea was to have a single base executor from which all executors would derive from. This would have allowed
all the common code to be shared among all the executors. The base executor had a CRTP based design that
was used to access the properties of the derived executor. It also allowed simplifications in many areas, such as
being able to reference any derived executor. Basic properties of all executors like
copy constructors, overloaded equality operators. The CRTP mechanism had some restrictions in the sense that
there was still a need for templates, and there were concerns regarding explicitly passing all the properties
to the base class using CRTP as it was felt to be unnecessary. Having inheritance also introduces runtime polymorphism
as the derived executor would override some of the base executors' method, and the call to the overridden function is
resolved at run time. This is opposed to one of the design consideration i.e., have everything compile-time and avoid

> any overhead. The base executor didn't add a lot besides allowing code sharing of a few common functionalities and

Reviewer: ...and avoid runtime overhead. 

You still pay the price in compilation time. 
Something to watch in case you're feeling lazy at some point:
https://www.youtube.com/watch?v=rHIkrotSwcc

Reviewer 2: > Something to watch in case you're feeling lazy at some point:

Links to a 1 hour info-dense video by Chandler. I wonder how Sergio relaxes 😆
